### PR TITLE
Fix interoperability when MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE is disabled

### LINKED ISSUE #9551: Disabling MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE reduces interoperability
As of Mbed TLS 3.6.1, the documentation of  suggests that disabling it breaks TLS 1.3 when an incompatible middlebox is disabled. But it actually breaks interoperability with any TLS 1.3 client or server that has middlebox compatibility active, such as OpenSSL or GnuTLS with default settings.

~I will fix the documentation in https://github.com/Mbed-TLS/mbedtls/pull/9546 (+ forward port).~ Thanks to Ronald's advice, the fix is very easy, all it takes is to adapt the tests. Done in https://github.com/Mbed-TLS/mbedtls/pull/9563.

The definition of done for this issue is that disabling `MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE` only breaks interoperability when a middlebox is involved, and that it has no effect on interoperability with a peer that has middlebox compatibility enabled. This is how GnuTLS and OpenSSL behave. In particular, even when `MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE` is disabled, `ssl_client1` and `ssl_server` should interoperate with GnuTLS and OpenSSL with their default settings (without requiring `-no_middlebox` or `%DISABLE_TLS13_COMPAT_MODE`).

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
When `MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE` is disabled, if an Mbed TLS client or server attempted a handshake with a peer that uses compatibility mode, it would abort the handshake. Fix that: now, even with `MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE` disabled, Mbed TLS can communicate with e.g. a default OpenSSL or GnuTLS or Mbed TLS, as long as there is no middlebox in the way. Fixes #9551.

Content of this PR:

* The very small actual fix.
* Some massaging of ssl-opt test cases so that they are more suitable for automated processing.
* Remove `requires_config_enabled MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE` except when testing middlebox compatibility mode.
* Register `tests/opt-testcases/tls13-compat.sh` as a generated file.

Differences with https://github.com/Mbed-TLS/mbedtls/pull/9563:

* The commits "Avoid multiline requires_all_configs_xxx" and "Remove obsolete requirements on middlebox compatibility mode: manual" are done by running the same Perl script, rather than applying the same patch.
* Trivial conflict in "Move generation of tls13-compat.sh to tests/CMakeLists.txt" which adds content next to content that's been removed.
* An additional commit "Have `make ssl-opt` generate `tls13-compat.sh`" takes care of https://github.com/Mbed-TLS/mbedtls/pull/9565#discussion_r1767327667 since the other PR has just been merged. For 3.6, I'll do a follow-up PR (there's no harm now since none of the reference configurations enable TLS 1.3).

## PR checklist

- [x] **changelog** provided
- [x] **development PR** here
- [x] **framework PR** not required
- [x] **3.6 PR** https://github.com/Mbed-TLS/mbedtls/pull/9563
- [x] **2.28 PR** not required because: feature not in 2.28
- **tests**  provided

### CODE REVIEW

**Review by ronald-cron-arm — DISMISSED** [review]
I've been through the rebase of #9563 and ended up with the same tree as in this PR HEAD~1. The last commit looks good to me.

**Review by mpg — APPROVED** [review]
LGTM as a faithful forward port of #9563, with an additional commit at the end addressing https://github.com/Mbed-TLS/mbedtls/pull/9563#discussion_r1767328386

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -78,6 +78,7 @@ if(GEN_FILES)
     add_custom_target(tls13-compat.sh
         DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/opt-testcases/tls13-compat.sh)
     set_target_properties(tls13-compat.sh PROPERTIES EXCLUDE_FROM_ALL NO)
+    add_dependencies(${ssl_opt_target} tls13-compat.sh)
```
Comment (mpg) [inline tests/CMakeLists.txt:81]: Actually wait, shouldn't we do something similar for make as well?

Comment (gilles-peskine-arm) [reply tests/CMakeLists.txt:81]: Good point! I'd only thought about it from the point of view of having already run `make generated_files`, which doesn't exist with CMake. But I do expect all make targets to work from a fresh checkout. Fixed.

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Review by ronald-cron-arm — APPROVED** [review]

**Pull request merged**